// @pwabuilder/cli/dist/util/textUtil.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatSuccessEmphasis = exports.formatErrorEmphasisWeak = exports.formatErrorEmphasisStrong = exports.formatEmphasisStrong = exports.formatEmphasis = exports.formatCodeSnippet = void 0;
const chalk_1 = __importDefault(require("chalk"));
const lavender = chalk_1.default.hex('#cc99ff');
const purple = chalk_1.default.hex('#9966ff');
function formatCodeSnippet(text) {
    return lavender.bold(text);
}
exports.formatCodeSnippet = formatCodeSnippet;
function formatEmphasis(text) {
    return purple.bold(text);
}
exports.formatEmphasis = formatEmphasis;
function formatEmphasisStrong(text) {
    return purple.inverse(text);
}
exports.formatEmphasisStrong = formatEmphasisStrong;
function formatErrorEmphasisStrong(text) {
    return chalk_1.default.red.inverse(text);
}
exports.formatErrorEmphasisStrong = formatErrorEmphasisStrong;
function formatErrorEmphasisWeak(text) {
    return chalk_1.default.bold.red(text);
}
exports.formatErrorEmphasisWeak = formatErrorEmphasisWeak;
function formatSuccessEmphasis(text) {
    return chalk_1.default.greenBright(text);
}
exports.formatSuccessEmphasis = formatSuccessEmphasis;
