// @pwabuilder/cli/dist/commands/start.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.builder = exports.desc = exports.command = void 0;
const util_1 = require("../util/util");
const usage_analytics_1 = require("../analytics/usage-analytics");
const campaignUtil_1 = require("../util/campaignUtil");
const COMMAND_DESCRIPTION_STRING = 'Run the PWA Starter on a Vite dev server.';
const VITEARGS_DESCRIPTION_STRING = 'Arguments to pass directly to the Vite start process.';
exports.command = 'start';
exports.desc = COMMAND_DESCRIPTION_STRING;
const USAGE_STRING = '$0 start [viteArgs]';
const EXEC_START_NO_ARGS_STRING = 'npm run dev-server';
const EXEC_START_ARGS_STRING = (viteArgs) => { return 'npm run dev-task -- ' + viteArgs; };
const INVALID_DIRECTORY_ERROR_STRING = `Cannot execute start because the current working directory is not a valid PWA Starter template. 
Make sure you are executing the start command from a PWA Starter template directory.`;
const builder = (yargs) => yargs
    .options({
    viteArgs: { type: 'string', description: VITEARGS_DESCRIPTION_STRING }
})
    .usage(USAGE_STRING);
exports.builder = builder;
const handler = (argv) => {
    try {
        (0, campaignUtil_1.handleCampaign)(campaignUtil_1.WHISPER_CAMPAIGN);
        (0, usage_analytics_1.trackStartEventWrapper)();
        handleStartCommand(argv);
    }
    catch (error) {
        (0, usage_analytics_1.trackErrorWrapper)(error);
    }
};
exports.handler = handler;
function handleStartCommand(argv) {
    const { viteArgs } = argv;
    if ((0, util_1.isDirectoryTemplate)()) {
        spawnStartCommand(viteArgs);
    }
    else {
        (0, util_1.outputError)(INVALID_DIRECTORY_ERROR_STRING);
    }
}
function spawnStartCommand(viteArgs) {
    if (viteArgs) {
        (0, util_1.spawnWrapper)(EXEC_START_ARGS_STRING(viteArgs));
    }
    else {
        (0, util_1.spawnWrapper)(EXEC_START_NO_ARGS_STRING);
    }
}
